python/taichi/lang/_wrap_inspect.py

```diff
@@ -0,0 +1,143 @@
+    # Now we have found the filename and code lines.
+    # We first check if they are already cached, to avoid file io in each query.
+    try:
+        filename = _blender_findsource._saved_inspect_cache[lines]  # pylint: disable=no-member
+    except KeyError:
+        fd, filename = tempfile.mkstemp(prefix='_Blender_',
+                                        suffix=f'_{text_name}.py')
+        os.close(fd)
+
+        with open(filename, 'w') as f:
+            f.write(lines)
+
+        _blender_findsource._saved_inspect_cache[lines] = filename  # pylint: disable=no-member
+
+    def wrapped_getfile(ob):
+        if id(ob) == id(obj):
+            return filename
+
+        return _builtin_getfile(ob)
+
+    return _find_source_with_custom_getfile_func(wrapped_getfile, obj)
+
+
+_blender_findsource._saved_inspect_cache = {}
+
+
+def _Python_IPython_findsource(obj):
+    try:
+        # In Python and IPython the builtin findsource would suffice in most cases
+        return _builtin_findsource(obj)
+    except IOError:
+        # Except that the cell has a magic command like %%time or %%timeit
+        # In this case the filename returned by getfile is wrong
+        filename = _builtin_getfile(obj)
+        if (filename in {"<timed exec>", "<magic-timeit>"}):
+            try:
+                ip = get_ipython()
+                if ip is not None:
+                    session_id = ip.history_manager.get_last_session_id()
+                    fd, filename = tempfile.mkstemp(prefix='_IPython_',
+                                                    suffix=f'_{session_id}.py')
+                    os.close(fd)
+
+                    lines = ip.history_manager._i00
+                    # Remove the magic command (and spaces/sep around it) before saving to a file
+                    index = lines.find("%time")
+                    lines_stripped = lines[index:]
+                    lines_stripped = lines_stripped.split(maxsplit=1)[1]
+
+                    with open(filename, 'w') as f:
+                        f.write(lines_stripped)
+
+                    func = lambda obj: filename
+                    return _find_source_with_custom_getfile_func(func, obj)
+
+            except:
+                pass
+        raise IOError(f"Cannot find source code for Object: {obj}")
+
+
+def _custom_findsource(obj):
+    try:
+        return _Python_IPython_findsource(obj)
+    except IOError:
+        try:
+            return _blender_findsource(obj)
+        except:
+            raise IOError(f"Cannot find source code for Object: {obj} ")
+
+
+class _InspectContextManager:
+    def __enter__(self):
+        inspect.findsource = _custom_findsource
+        return self
+
+    def __exit__(self, *_):
+        inspect.findsource = _builtin_findsource
+
+
+def _getsourcelines(obj):
+    try:
+        with _InspectContextManager():
+            return inspect.getsourcelines(obj)
+    except:
+        warn(f'Using sourceinspect to get source lines of {obj}')
```

Comment:
If we just automatically fallback to without asking users to open an issue, we'll probably never know when a bug happens and it'll bite us again when we remove source inspect.

Reply:
I'm afraid we cannot add a test here: the warning should only be raised in IDLE or Blender's interactive interpreter console, not the environment we can reproduce in a test?

Reply:
agh right, I guess my point was more like we should add a global toggle(maybe env variable?)that when it's turned on, everything just falls back to source inspect. I think this is also what we agreed on last time right? (having a way for users to fallback instead of fallback for them automatically?

Reply:
I see. Shall we make it an option in `ti.init()`?

Reply:
I guess an env is much simpler?